here-naksha-app-service/src/main/java/com/here/naksha/app/service/http/tasks/WriteFeatureApiTask.java

```diff
@@ -228,6 +238,202 @@ protected void init() {}
     }
   }
 
+  // TODO this might change if naksha_plpgsql.sql gets updated and the message is modified
+  private static final Pattern FEATURE_ID_FROM_ERR = Pattern.compile("The feature \\s*'([^']*)' uuid");
+  private static final int MAX_RETRY_ATTEMPT = 5;
+
+  private @NotNull XyzResponse executePatchFeatureById() throws JsonProcessingException {
+
+    final XyzFeature featureFromRequest = singleFeatureFromRequestBody();
```

Comment:
IntelliJ rightly reports that this code fragment is duplicated - take a look at `executeUpdateFeature`. Both of these method delegate this common fragment of code to some dedicated private method. 
![Uploading image.png…]()

Reply:
This is interestingly the direct counterpart of the problem below :) 

On the one hand here, we have reusable methods like `extractMandatoryPathParam()`, `extractParamAsStringList()`, that can be reusable, and also left out whenever not applicable (like not all need the added and removed tags). But since this code fragment appear twice, it should be grouped into a shared method. This kind of defeats the purpose of `singleFeatureFromRequestBody()`, since so far this has only been used at these 2 places.

On the other hand, below we have an indeed very huge chunk of code below, which is quite coupled and is planned to be reused at only 2 places, but should be abstracted out into further smaller reusable methods.

Personally I say the sweet spot for deciding when to make a reusable block is that this will potentially be used at 4 different places (or in the case below because it's recursive). Abstracting more than this will make us keep toggling around different methods in different classes while reading the code, which unfortunately reduces clarity instead.

Maybe it's worth discussing this sometimes, maybe with the other guys in the technical chat as well. But I will push the change for this shared block soon.
Some thread I found: https://stackoverflow.com/questions/2668355/how-much-abstraction-is-too-much